subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiClientJdkCompatibilityTest.groovy

```diff
@@ -26,6 +26,12 @@ abstract class ToolingApiClientJdkCompatibilityTest extends AbstractIntegrationS
     def setup() {
         System.out.println("TAPI client is using Java " + clientJdkVersion)
 
+        def jvmArgs = """
+            if (${clientJdkVersion.isCompatibleWith(JavaVersion.VERSION_16)} && ['2.6', '2.14.1'].contains(project.findProperty("gradleVersion"))) {
+                jvmArgs = ["--add-opens", "java.base/java.lang=ALL-UNNAMED"]
```

Comment:
❌ Can we set these parameters automatically when the TAPI client targets those Gradle versions?

Reply:
If I understand this test correctly, it will launch a Java application using JavaExec using a given JDK (16 in this case). It is that Java application (`ToolingApiCompatibilityClient.main`) that depends on the latest TAPI and connects to an older version of TAPI. This connection seems to happen within the same JVM.
The required `--add-opens` here need to be set at the JVM start and because there is no forking happening in between TAPI current -> TAPI XX, I don't think we can automate this.

Because within TAPI current we know which JVM we're on and which version we are targeting, we could either fail gracefully and tell the user that this combination is not supported, or alternatively, fail gracefully and ask the user to pass the required `--add-opens` to the JVM. Although on the other hand, the latter option is what happens anyway, just with a stack trace. Thinking of it more, I think the second option is not a clean one as it would require some possibly dirty code to assert that the reflective access is allowed and fail only then.

I'd lean to either explicitly saying "sorry, please upgrade TAPI to at least XX (3.x or 4.x)" for JDK16 compatibility, or leaving this as-is, allowing users to read the stack trace and get out of the situation by passig the required arguments to the JVM.

What do you think?

Reply:
Thanks for the detailed explanation. Now I understand what's going on. Basically, we need the special JVM arguments to support older TAPI versions. It's fine to have those `--add-opens` for this case.